### PR TITLE
Kalinin Alexandr Lab 3 Var 2

### PR DESCRIPTION
**Pull request by KalininA1**

### CODE REVIEW

File: llvm/lib/Target/X86/lab3/kalinin_alexandr/X86KalininMICounter.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      LLVMContext &llvmContext = module.getContext();
+      globalCounter =
+          new GlobalVariable(module, IntegerType::get(llvmContext, 64), false,
+                             GlobalValue::ExternalLinkage, nullptr, "ic");
+    }
+
+    for (auto &basicBlock : machineFunction) {
+      int instructionCount =
+          std::distance(basicBlock.begin(), basicBlock.end());
+      auto insertPoint = basicBlock.getFirstTerminator();
```
Comment (m-ly4) [inline llvm/lib/Target/X86/lab3/kalinin_alexandr/X86KalininMICounter.cpp]: Please insert instruction into a first valid location in a block.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/X86/lab3/kalinin_alexandr/X86KalininMICounter.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      globalCounter =
+          new GlobalVariable(module, IntegerType::get(llvmContext, 64), false,
+                             GlobalValue::ExternalLinkage, nullptr, "ic");
+    }
+
+    for (auto &basicBlock : machineFunction) {
+      size_t instructionCount =
+          std::distance(basicBlock.begin(), basicBlock.end());
+
+      BuildMI(basicBlock, basicBlock.getFirstTerminator(), debugLoc,
```
Comment (m-ly4) [inline llvm/lib/Target/X86/lab3/kalinin_alexandr/X86KalininMICounter.cpp]: Your previous version was more correct. All I asked was to insert Add instruction into the beginning of a basic block, but taking into account that some instructions should be left the very first.

File: llvm/test/lab3/kalinin_alexandr/test.mir [hunk 3]
```diff
@@ -0,0 +1,252 @@
+# RUN: llc -mtriple x86_64-unknown-linux-gnu --load=%llvmshlibdir/X86SKalininMICounterPass%shlibext -run-pass=x86-kalinin-mi-counter  %s -o - | FileCheck %s
+
+# test.cpp
+#
+# unsigned long ic;
```
Comment (aobolensk) [inline llvm/test/lab3/kalinin_alexandr/test.mir]: Add check for the case when there is no "ic" to cover the situation when you create "ic" by yourself

**Review by aobolensk — CHANGES_REQUESTED** [review]

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/lab3/kalinin_alexandr/test.mir (hunk 3 above)
Comment (KalininA1) [reply llvm/test/lab3/kalinin_alexandr/test.mir]: Sorry, but Mikhail Mecheslavovich said that according to the conditions of the problem, the variable ic exists. Still, do we need to display this case too?

Comment (aobolensk) [reply llvm/test/lab3/kalinin_alexandr/test.mir]: In this case there is no need to create global variable in the pass

Comment (KalininA1) [reply llvm/test/lab3/kalinin_alexandr/test.mir]: Sorry, as a result, I need to redo the tests so that there is no ic variable?

Comment (aobolensk) [reply llvm/test/lab3/kalinin_alexandr/test.mir]: No-no. Just need to remove if statement where you create "ic" global var. That's it

File: llvm/lib/Target/X86/lab3/kalinin_alexandr/X86KalininMICounter.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      while (insertPos != basicBlock.end() &&
+             (insertPos->isPHI() || insertPos->isDebugValue())) {
+        ++insertPos;
+      }
+
+      size_t instructionCount =
+          std::distance(basicBlock.begin(), basicBlock.end());
+
+      BuildMI(basicBlock, insertPos, debugLoc,
+              targetInstrInfo->get(X86::ADD64ri32))
```
Comment (m-ly4) [inline llvm/lib/Target/X86/lab3/kalinin_alexandr/X86KalininMICounter.cpp]: This instruction is for register input, whilst `ic` is address in memory.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/lab3/kalinin_alexandr/test.mir (hunk 3 above)
Comment (KalininA1) [reply llvm/test/lab3/kalinin_alexandr/test.mir]: ok, I understand you

**Review by aobolensk — APPROVED** [review]
Can be considered accepted after @m-ly4 's approval and CI pass

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**